Library/Homebrew/formula.rb

```diff
@@ -980,25 +998,45 @@ def migration_needed?
   end
 
   # @private
-  def outdated_versions
-    @outdated_versions ||= begin
-      all_versions = []
-
+  def outdated_versions(options = {})
+    begin
       raise Migrator::MigrationNeededError.new(self) if migration_needed?
 
-      installed_kegs.each do |keg|
-        version = keg.version
-        all_versions << version
-        return [] if pkg_version <= version
+      if @outdated_versions && !options[:fetch_head]
+        return @outdated_versions
+      elsif @outdated_versions_head_fetched && options[:fetch_head]
+        return @outdated_versions_head_fetched
       end
 
+      if options[:fetch_head]
+        @outdated_versions_head_fetched ||= _outdated_versions(options)
+      else
+        @outdated_versions ||= _outdated_versions(options)
+      end
```

Comment:
Instead of introducing two variable and some complicated logics, you can actually do this:
```ruby
def outdated_versions(options = {})
  @outdated_versions ||= Hash.new do |options|
    @outdated_versions[options] = _outdated_versions(options)
  end
  @outdated_versions[options]
end
```